.github/workflows/ci.yml

```diff
@@ -53,61 +53,72 @@ jobs:
             substituters = https://nix-cache.lowrisc.org/public/ https://cache.nixos.org/
             trusted-public-keys = nix-cache.lowrisc.org-public-1:O6JLD0yXzaJDPiQW1meVu32JIDViuaPtGDfjlOopU7o= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
 
-      # Attic is not in nixpkgs currently, so it's not in NixOS cache. In case it's also not in our cache,
-      # we want to avoid having to build it for every job. Do a build ahead of time and cache attic itself.
-      - name: Build and Cache Attic
-        if: github.event_name != 'pull_request'
-        run: |
-          nix profile install github:zhaofengli/attic#attic
-          attic login --set-default lowrisc https://nix-cache.lowrisc.org/ ${{ secrets.NIX_CACHE_TOKEN }}
-          attic push public $(which attic)
-
       - name: Generate Build Plan
         id: plan
         run: |
           REBUILD=()
-          echo '| Package | Outputs | Status |' >> $GITHUB_STEP_SUMMARY
-          echo '|---------|---------|--------|' >> $GITHUB_STEP_SUMMARY
-
-          for LINE in $(nix-shell -p nix-eval-jobs --run 'nix-eval-jobs --accept-flake-config --flake .#packages.x86_64-linux'); do
-            ATTR=$(echo "$LINE" | jq -r .attr)
-            OUTPUTS=$(echo "$LINE" | jq -r '.outputs | keys | join(",")')
-
-            # Check if all outputs are cached. We don't use nix-eval-jobs's cache check because it also includes local path
-            CACHED=y
-            for OUTPUT in ${OUTPUTS//,/ }; do
-              HASH=$(echo "$LINE" | jq -r ".outputs.$OUTPUT" | cut -d '/' -f4 | cut -d '-' -f1)
-              if ! curl -sSfL "https://nix-cache.lowrisc.org/public/$HASH.narinfo"; then
-                CACHED=n
+          echo '| Package | System | Status |' >> $GITHUB_STEP_SUMMARY
+          echo '|---------|--------|--------|' >> $GITHUB_STEP_SUMMARY
+
+          if ${{ github.event_name == 'pull_request' }}; then
+            # We only support MacOS with best-effort basis. They take much longer to build so shouldn't block PR.
+            SYSTEMS=(x86_64-linux)
+          else
+            SYSTEMS=(x86_64-linux x86_64-darwin aarch64-darwin)
+          fi
+
+          for SYSTEM in "${SYSTEMS[@]}"; do
+            # Choose a GitHub runner for the given system.
+            case $SYSTEM in
+              x86_64-linux)
+                RUNNER=nixos-23.11
+                ;;
+              x86_64-darwin)
+                RUNNER=macos-13
+                ;;
+              aarch64-darwin)
+                RUNNER=macos-14
+                ;;
+            esac
+
+            # This is a json object with package names as keys and their main outputs as values
+            PACKAGES=$(nix eval --json ".#packages.$SYSTEM")
+
+            for ATTR in $(echo "$PACKAGES" | jq -r 'keys[]'); do
+              HASH=$(echo "$PACKAGES" | jq -r ".[\"$ATTR\"]" | cut -d '/' -f4 | cut -d '-' -f1)
+
+              # Check if the main output is cached by us or upstream.
```

Comment:
Is there a way to do this using the nix CLI?

Reply:
You can know if rebuilding is necessary, but there's no way to know if it's because it exists in local store or in one of the HTTP cache. Since we have self-hosted runners, we can still have cases where we need to push to the store even if it exists locally.

Reply:
Okay cool, it would be nice if in the future the tool could tell us this to avoid the .narinfo implementation detail.